common/flatpak-installation.c

```diff
@@ -1092,7 +1098,28 @@ flatpak_installation_list_installed_refs_for_update (FlatpakInstallation *self,
       /* Note: local_commit may be NULL here */
       if (remote_commit != NULL &&
           g_strcmp0 (remote_commit, local_commit) != 0)
-        g_ptr_array_add (updates, g_object_ref (installed_ref));
+        {
+          g_ptr_array_add (updates, g_object_ref (installed_ref));
+
+          /* Don't check further, as we already added the installed_ref to @updates. */
+          continue;
+        }
+
+      /* Check if all "should-download" related refs for the ref are installed.
+       * If not, add the ref in @updates array so that it can be installed via
+       * FlatpakTransaction's update-op.
+       *
+       * This makes sure that the ref (maybe an app or runtime) remains in usable
+       * state and fixes itself through an update.
+       */
+      state = flatpak_dir_get_remote_state_optional (dir, remote_name, FALSE, cancellable, error);
```

Comment:
Hmm, this still does the get_remote_state for each installed ref, rather than once per remote. But, i see now that this happens not inside the iteration over all remotes, but rather in a different loop. We could still avoid doing it by caching the states in a per-remote hashtable during the entire function. But, lets leave that for a followup so we can land this.

Reply:
This was resolved by https://github.com/flatpak/flatpak/pull/3259